docs(FEC-9468): ad layout waterfall documentation (#289)
kaltura/kaltura-player-js: docs/advertising.md

# Ad Layout and Waterfalling

> Ad layout is a new config option lets the application to sets a scheme of vasts (instead of VMAP) with waterfalling mechanism.

Reviewer: In general, writing in docs about something being new will soon become obsolete, this is more appropriate to blog post that has a timestamp.

I would change the title to say 

> Advertisement layout management

And then the description body will say:

> With ad layout config you can create your own ad break timeline using your vast tags.
> Ad break can be set as pre, mid and post rolls and each ad break can contain a single vast tag or multiple tags, either as a pod, but also as a Waterfall*

And then add a short note saying "see what is waterfall in waterfall section" and add a short explainer there(maybe a task for Ron L)

## Table of Contents

- [Basic Usage](#basic-usage)
  - [Single Ad](#single-ad)
  - [Ad Pod](#ad-pod)

> - [Waterfalling](#waterfalling)

Reviewer: I think we can drop the basic usage section and flatten the TOC to
- Single Ad
- Ad Pod
- Waterfalling

### Basic Usage

#### Single Ad

Here's a simple scheme contains 4 ad breaks - 1 pre-roll, 2 mid-rolls and 1 post-roll:

```ecmascript 6
const config = {
  ...
  advertising: {
    adBreaks: [{
      position: 0,
      ads: [{
          url: [PRE_ROLL_VAST_URL]
        }]
      }, {
      position: 15,
      ads: [{
          url: [MID_ROLL_VAST_URL]
        }]
      }, {
      position: 60,
      ads: [{
          url: [MID_ROLL_VAST_URL]
        }]
      }, {
      position: -1,
      ads: [{
          url: [POST_ROLL_VAST_URL]
        }]
      }
    ]
  }
  ...
}
const kalturaPlayer = KalturaPlayer.setup(config);
```

> > Note: `position: 0` means pre-roll. `position: -1` means postroll.

Reviewer: do we expose the ad position type enum?

> > In this sample while the playback will be played 4 ads (Unless ad failures).

Reviewer: ```suggestion
> In this sample the player will try to request and play 4 ads
```

#### Ad Pod

> An ad break can contain not only a single ad, but also a list of ads (aka ad pod).

Reviewer: ```suggestion
An ad break may contain a list of ads, also referred to as ad pod.
IAB definition of ad pod:

> An individual ad pod is a group of ads expected to play back-to-back in one commercial ad break similar to how consumers experience commercial ad breaks in broadcast television. An ad pod can be of varying lengths and can be inserted at any point in a stream of content (pre, mid, or post).

```

```ecmascript 6
const config = {
  ...
  advertising: {
    adBreaks: [{
      position: 0,
      ads: [{
          url: [PRE_ROLL_1_VAST_URL]
        },{
          url: [PRE_ROLL_2_VAST_URL]
        }]
      }, {
      position: 15,
      ads: [{
          url: [MID_ROLL_1_VAST_URL]
        },{
          url: [MID_ROLL_2_VAST_URL]
        }]
      }, {
      position: 60,
      ads: [{
          url: [MID_ROLL_3_VAST_URL]
        },{
          url: [MID_ROLL_4_VAST_URL]
        }]
      }, {
      position: -1,
      ads: [{
          url: [POST_ROLL_1_VAST_URL]
        },{
          url: [POST_ROLL_2_VAST_URL]
        }]
      }
    ]
  }
  ...
}
const kalturaPlayer = KalturaPlayer.setup(config);
```

> > In this sample while the playback will be played 8 ads.

Reviewer: ```suggestion
> In this sample the player will try to request and play 8 ads in 4 ad breaks, each one containing 2 ads consecutively
```

### Waterfalling

An application may want to set a fallback vast url, in case the primary ad request will failed the ad won't be skipped.
This mechanism called _Waterfalling_, and it's configurable easily using the ad layout config.

> Here's the sample above with waterfalling:

Reviewer: Add the waterfall explainer here - so we start with answering the question "What is a waterfall?" - suggest to consult with Ron L on wording.
Then go on explaining how it is accomplished in player - "set an array of ad tags in a cascading order. The player will try to play the first one and if it fails will continue to the next until and ad request is successful or all ads have failed to load."

```ecmascript 6
const config = {
  ...
  advertising: {
    adBreaks: [{
      position: 0,
      ads: [{
          url: [PRE_ROLL_1.1_VAST_URL, PRE_ROLL_1.2_VAST_URL, ...]
        },{
          url: [PRE_ROLL_2.1_VAST_URL, PRE_ROLL_2.2_VAST_URL, ...]
        }]
      }, {
      position: 15,
      ads: [{
          url: [MID_ROLL_1.1_VAST_URL, MID_ROLL_1.2_VAST_URL, ...]
        },{
          url: [MID_ROLL_2.1_VAST_URL, MID_ROLL_2.2_VAST_URL, ...]
        }]
      }, {
      position: 60,
      ads: [{
          url: [MID_ROLL_3.1_VAST_URL, MID_ROLL_3.2_VAST_URL, ...]
        },{
          url: [MID_ROLL_4.1_VAST_URL, MID_ROLL_4.2_VAST_URL, ...]
        }]
      }, {
      position: -1,
      ads: [{
          url: [POST_ROLL_1.1_VAST_URL, POST_ROLL_1.2_VAST_URL, ...]
        },{
          url: [POST_ROLL_2.1_VAST_URL, POST_ROLL_2.2_VAST_URL, ...]
        }]
      }
    ]

>   }

Reviewer: Is the sample mixing both pod and waterfall, I think this is a great sample, but I would separate this sample to only waterfall and add only one waterfall in, for example, a midroll position.
I would than add another section, a forth one, that explained about the reach possibilities of mixing all formats in a timeline, i.e. set single vast preroll, 3 midrolls, one as a pod and one as a waterfall and another as a pod with waterfalling

This will cover all permutations and show how reach and powerful the feature is.

  ...
}
const kalturaPlayer = KalturaPlayer.setup(config);
```

> Important: A fallback url (e.g. PRE_ROLL_1.2_VAST_URL) will be only used if the previous request (PRE_ROLL_1.1_VAST_UR) is failed.
> Hence, In this sample while the playback will still be played 8 ads.

> > Note: There is no limit to the fallback url list.

Reviewer: add a note that each fallback url will result in longer time to ad loading in case all ads before it have failed.

Reviewer: did you add the note?
